File: latest/week1/workshop2_m8/W02_M08_demo.R
# ============================================================
# Workshop 2 — Module 8: labs(), Annotations & Storytelling
# R Demonstration Script — UNYT Tirana
# ============================================================
library(tidyverse)
library(patchwork)
dir.create("output", showWarnings = FALSE)

# ── 1. LABS() — full label set ────────────────────────────
p_labs <- ggplot(mpg, aes(x = displ, y = hwy, color = factor(cyl))) +
  geom_point(alpha = 0.6, size = 2) +
  labs(
    title = "Fuel Economy Decreases with Engine Size",
    subtitle = "Highway MPG vs displacement, coloured by cylinder count",
    caption = "Source: mpg dataset (ggplot2) | UNYT Data Visualization Course",
    x = "Displacement (litres)",
    y = "Highway MPG",
    color = "Cylinders"
  ) +
  scale_color_brewer(palette = "Set2") +
  theme_minimal(base_size = 11) +
  theme(
    plot.title = element_text(face = "bold", size = 14),
    plot.subtitle = element_text(color = "grey50"),
    plot.caption = element_text(face = "italic", hjust = 0, size = 8)
  )
ggsave("output/labs_full.png", p_labs, width = 7, height = 5, dpi = 300)

# ── 2. ANNOTATION TYPES ──────────────────────────────────
set.seed(33)
ts <- tibble(month = 1:12, revenue = cumsum(rnorm(12, 5, 8)) + 200)
mean_rev <- mean(ts$revenue)
peak <- ts |> slice_max(revenue, n = 1)

p_annotated <- ggplot(ts, aes(x = month, y = revenue)) +
  geom_line(color = "#1565C0", linewidth = 1.2) +
  geom_point(color = "#1565C0", size = 2) +
  # Reference line (mean)
  geom_hline(yintercept = mean_rev, linetype = "dashed", color = "grey50") +
  annotate("text", x = 12.5, y = mean_rev, label = paste0("Mean: $", round(mean_rev), "K"),
           size = 3, color = "grey50", fontface = "bold", hjust = 0) +
  # Target band
  annotate("rect", xmin = -Inf, xmax = Inf, ymin = 210, ymax = 230,
           alpha = 0.08, fill = "#2E7D32") +
  # Event marker
  geom_vline(xintercept = 6.5, linetype = "dotted", color = "#E65100") +
  annotate("text", x = 7, y = min(ts$revenue) - 3, label = "H1 | H2",
           size = 3, color = "#E65100") +
  # Peak callout
  annotate("text", x = peak$month + 1.5, y = peak$revenue + 10,
           label = paste0("Peak: $", round(peak$revenue), "K"),
           fontface = "bold", color = "#E53935", size = 3.5) +
  annotate("segment", x = peak$month + 1, y = peak$revenue + 8,
           xend = peak$month + 0.2, yend = peak$revenue + 1,
           arrow = arrow(length = unit(0.15, "cm")), color = "#E53935") +
  theme_minimal() +
  labs(title = "Monthly Revenue with Annotations",
       subtitle = "All four annotation types: text, hline, band, vline",
       x = "Month", y = "Revenue ($K)")

ggsave("output/annotations.png", p_annotated, width = 8, height = 5, dpi = 300)

# ── 3. STORYTELLING LAYERS — progressive build ───────────
set.seed(77)
ts2 <- tibble(year = 2015:2024,
              revenue = c(120, 135, 128, 142, 155, 170, 145, 180, 195, 220))

p1 <- ggplot(ts2, aes(x = year, y = revenue)) + geom_line(color = "#1565C0", linewidth = 1.2) +
  geom_point(color = "#1565C0", size = 2) + theme_minimal() + labs(title = "1. Show the data")

p2 <- p1 + geom_smooth(method = "lm", se = FALSE, color = "grey50", linewidth = 0.8) +
  labs(title = "2. + Trend line")

p3 <- p2 + geom_vline(xintercept = 2020, linetype = "dotted", color = "#E53935") +
  annotate("text", x = 2020.3, y = 155, label = "COVID", color = "#E53935",
           fontface = "bold", size = 3) +
  labs(title = "3. + Key events")

p4 <- p3 + labs(title = "Revenue Recovered to Record $220M in 2024",
                subtitle = "Annual revenue with COVID impact and recovery trend")

p_story <- (p1 | p2) / (p3 | p4) +
  plot_annotation(title = "Storytelling Layers: Progressive Reveal")
ggsave("output/storytelling.png", p_story, width = 10, height = 7, dpi = 300)

# ── 4. ANNOTATED BAR — complete example ──────────────────
df <- mpg |> count(class) |> mutate(class = fct_reorder(class, n))
mean_n <- mean(df$n)

p_final <- df |>
  mutate(highlight = class == "suv") |>
  ggplot(aes(x = class, y = n, fill = highlight)) +
  geom_col(width = 0.5, show.legend = FALSE) +
  geom_text(aes(label = n), hjust = -0.2, size = 3.5, fontface = "bold") +
  scale_fill_manual(values = c("TRUE" = "#E53935", "FALSE" = "#BBBBBB")) +
  geom_hline(yintercept = mean_n, linetype = "dashed", color = "grey50") +
  coord_flip() +
  labs(title = "SUV Leads with 62 Models",
       subtitle = "Vehicle count by class, mpg dataset",
       caption = "Source: ggplot2::mpg | UNYT Data Visualization Course",
       x = NULL, y = NULL) +
  theme_minimal(base_size = 11) +
  theme(
    plot.title = element_text(face = "bold", size = 14),
    plot.subtitle = element_text(color = "grey50"),
    plot.caption = element_text(face = "italic", hjust = 0),
    panel.grid.major.y = element_blank()
  )
ggsave("output/annotated_bar.png", p_final, width = 7, height = 5, dpi = 300)
ggsave("output/annotated_bar.pdf", p_final, width = 7, height = 5)

cat("\n── All W02-M08 R plots saved ──\n")

File: latest/week1/workshop2_m8/W02_M08_demo.py
"""W02-M08: labs(), Annotations & Storytelling — Python — UNYT Tirana"""
import numpy as np, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True); np.random.seed(42)

# 1. ANNOTATION TYPES — all four on one chart
np.random.seed(33); months=np.arange(1,13); revenue=np.cumsum(np.random.normal(5,8,12))+200
mean_rev=np.mean(revenue); peak_idx=np.argmax(revenue)

fig,ax=plt.subplots(figsize=(8,5))
ax.plot(months,revenue,"-o",color="#1565C0",linewidth=2,markersize=6)
# Reference line
ax.axhline(y=mean_rev,color="grey",lw=1,ls="--")
ax.text(12.5,mean_rev,f"Mean: ${mean_rev:.0f}K",fontsize=7,color="grey",fontweight="bold",va="center")
# Target band
ax.axhspan(210,230,alpha=0.08,color="#2E7D32")
ax.text(0.5,220,"Target",fontsize=7,color="#2E7D32",fontweight="bold")
# Event marker
ax.axvline(x=6.5,color="#E65100",lw=0.8,ls=":")
ax.text(6.7,min(revenue)-3,"H1 | H2",fontsize=7,color="#E65100")
# Peak callout
ax.annotate(f"Peak: ${revenue[peak_idx]:.0f}K",xy=(months[peak_idx],revenue[peak_idx]),
    xytext=(months[peak_idx]+1.5,revenue[peak_idx]+15),fontsize=9,fontweight="bold",color="#E53935",
    arrowprops=dict(arrowstyle="->",color="#E53935",lw=1.5))
ax.set_xlabel("Month"); ax.set_ylabel("Revenue ($K)")
ax.set_title("Four Annotation Types on One Chart",fontsize=12,fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.tick_params(labelsize=7); ax.grid(alpha=0.1)
plt.tight_layout(); plt.savefig("output/annotations_py.png",dpi=300); plt.close()

# 2. STORYTELLING LAYERS
np.random.seed(77); years=np.arange(2015,2025); rev2=np.array([120,135,128,142,155,170,145,180,195,220])
fig,axes=plt.subplots(2,2,figsize=(10,7))
titles=["1. Show the data","2. + Trend line","3. + Key events","4. + Title = finding"]
for i,ax in enumerate(axes.flatten()):
    ax.plot(years,rev2,"-o",color="#1565C0",linewidth=1.5,markersize=4)
    if i>=1:
        m,b=np.polyfit(years,rev2,1); xs=np.linspace(2015,2024,50)
        ax.plot(xs,m*xs+b,color="grey",linewidth=1,linestyle="--")
    if i>=2:
        ax.axvline(x=2020,color="#E53935",linewidth=1,linestyle=":")
        ax.text(2020.2,150,"COVID",fontsize=7,color="#E53935",fontweight="bold")
    if i>=3:
        ax.set_title("Revenue Recovered to\nRecord $220M in 2024",fontsize=9,fontweight="bold",color="#1565C0")
    else:
        ax.set_title(titles[i],fontsize=9,fontweight="bold")
    ax.tick_params(labelsize=6); ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Storytelling Layers: Progressive Reveal",fontsize=12,fontweight="bold")
plt.tight_layout(); plt.savefig("output/storytelling_py.png",dpi=300); plt.close()

# 3. ANNOTATED BAR — complete example
cats=["SUV","Pickup","Midsize","Compact","Minivan"]; vals=[62,33,41,47,11]
idx=np.argsort(vals); sc=[cats[i] for i in idx]; sv=[vals[i] for i in idx]
fig,ax=plt.subplots(figsize=(6,5))
colors=["#BBBBBB"]*5; colors[-1]="#E53935"
ax.barh(sc,sv,color=colors,height=0.5,edgecolor="none")
for i,v in enumerate(sv):
    ax.text(v+0.8,i,str(v),va="center",fontsize=9,fontweight="bold",color="#E53935" if i==4 else "#888")
ax.axvline(x=np.mean(sv),color="#888",lw=0.8,ls="--")
ax.set_title("SUV Leads with 62 Models",fontsize=12,fontweight="bold")
fig.text(0.15,0.90,"Vehicle count by class, 2024",fontsize=9,color="grey")
fig.text(0.15,0.02,"Source: mpg | UNYT",fontsize=7,fontstyle="italic",color="#AAA")
for sp in ["top","right","left"]: ax.spines[sp].set_visible(False)
ax.set_xticks([]); ax.tick_params(left=False,labelsize=9)
plt.subplots_adjust(top=0.84,bottom=0.08)
plt.savefig("output/annotated_bar_py.pdf",bbox_inches="tight")
plt.savefig("output/annotated_bar_py.png",dpi=300,bbox_inches="tight"); plt.close()

print("All W02-M08 Python plots saved")
